File: csharp/tests/EnvoiceTests.cs
using Xunit;
using Envoice;

namespace Envoice.Tests;

public class EnvoiceClientTests
{
    [Fact]
    public void Client_RequiresApiKey()
    {
        Assert.Throws<ArgumentException>(() => new EnvoiceClient(""));
        Assert.Throws<ArgumentException>(() => new EnvoiceClient("   "));
    }

    [Fact]
    public void Client_CreatesWithValidApiKey()
    {
        using var client = new EnvoiceClient("env_sandbox_test");
        Assert.NotNull(client);
    }

    [Fact]
    public void Invoice_ReturnsBuilder()
    {
        using var client = new EnvoiceClient("env_sandbox_test");
        var builder = client.Invoice();
        Assert.NotNull(builder);
        Assert.IsType<InvoiceBuilder>(builder);
    }
}

public class InvoiceBuilderTests
{
    [Fact]
    public async Task Builder_ValidatesRequiredFields()
    {
        using var client = new EnvoiceClient("env_sandbox_test");
        var result = await client.Invoice().GenerateAsync();

        Assert.False(result.IsSuccess);
        Assert.IsType<InvoiceFailure>(result);

        var failure = (InvoiceFailure)result;
        Assert.Contains(failure.Errors, e => e.Path == "$.invoice.number");
        Assert.Contains(failure.Errors, e => e.Path == "$.invoice.date");
        Assert.Contains(failure.Errors, e => e.Path == "$.invoice.seller");
        Assert.Contains(failure.Errors, e => e.Path == "$.invoice.buyer");
        Assert.Contains(failure.Errors, e => e.Path == "$.invoice.items");
    }

    [Fact]
    public void Builder_FluentInterface()
    {
        using var client = new EnvoiceClient("env_sandbox_test");
        var builder = client.Invoice();

        Assert.Same(builder, builder.Number("2026-001"));
        Assert.Same(builder, builder.Date("2026-01-15"));
        Assert.Same(builder, builder.DueDate("2026-02-15"));
        Assert.Same(builder, builder.Currency("EUR"));
        Assert.Same(builder, builder.Template("minimal"));
        Assert.Same(builder, builder.Locale("de"));
        Assert.Same(builder, builder.FooterText("Thanks!"));
        Assert.Same(builder, builder.AccentColor("#8b5cf6"));
    }

    [Fact]
    public void Builder_AcceptsPartyBuilders()
    {
        using var client = new EnvoiceClient("env_sandbox_test");

        var builder = client.Invoice()
            .Seller(s => s.Name("Acme GmbH").VatId("DE123456789").City("Berlin").Country("DE"))
            .Buyer(b => b.Name("Customer AG").City("München").Country("DE"));

        Assert.NotNull(builder);
    }

    [Fact]
    public void Builder_AcceptsLineItemBuilders()
    {
        using var client = new EnvoiceClient("env_sandbox_test");

        var builder = client.Invoice()
            .AddItem(i => i.Description("Consulting").Quantity(8).Unit("HUR").UnitPrice(150).VatRate(19));

        Assert.NotNull(builder);
    }
}

public class InvoiceResultTests
{
    [Fact]
    public void Success_ProvidesCorrectData()
    {
        var success = new InvoiceSuccess(
            pdfBase64: "JVBERi0xLjQKJeLjz9MKMSAwIG9iago8PC9UeXBlL0NhdGFsb2c+PgplbmRvYmoKdHJhaWxlcgo8PC9Sb290IDEgMCBSPj4KJSVFT0YK",
            filename: "invoice-2026-001.pdf",
            validation: new ValidationResult("valid", "EN16931", "2.3.2"),
            account: new AccountInfo(499, "starter")
        );

        Assert.True(success.IsSuccess);
        Assert.Equal("invoice-2026-001.pdf", success.Filename);
        Assert.Equal("EN16931", success.Validation.Profile);
        Assert.Equal(499, success.Account?.Remaining);

        var bytes = success.ToBytes();
        Assert.NotEmpty(bytes);
        Assert.StartsWith("%PDF", System.Text.Encoding.UTF8.GetString(bytes[..4]));

        var dataUrl = success.ToDataUrl();
        Assert.StartsWith("data:application/pdf;base64,", dataUrl);
    }

    [Fact]
    public void Failure_ContainsErrors()
    {
        var failure = new InvoiceFailure(new List<ValidationError>
        {
            new("$.invoice.number", "REQUIRED", "Invoice number is required")
        });

        Assert.False(failure.IsSuccess);
        Assert.True(failure.IsFailure);
        Assert.Single(failure.Errors);
        Assert.Equal("$.invoice.number", failure.Errors[0].Path);
    }
}

File: swift/Tests/EnvoiceTests/EnvoiceTests.swift
import XCTest
@testable import Envoice

final class EnvoiceTests: XCTestCase {

    func testClientRequiresApiKey() {
        XCTAssertThrowsError(try EnvoiceClient(apiKey: "")) { error in
            XCTAssertEqual(error as? EnvoiceError, EnvoiceError.invalidApiKey)
        }
    }

    func testClientCreatesWithValidApiKey() throws {
        let client = try EnvoiceClient(apiKey: "env_sandbox_test")
        XCTAssertNotNil(client)
    }

    func testInvoiceReturnsBuilder() throws {
        let client = try EnvoiceClient(apiKey: "env_sandbox_test")
        let builder = client.invoice()
        XCTAssertNotNil(builder)
    }

    func testBuilderFluentInterface() throws {
        let client = try EnvoiceClient(apiKey: "env_sandbox_test")
        let builder = client.invoice()

        XCTAssertTrue(builder === builder.number("2026-001"))
        XCTAssertTrue(builder === builder.date("2026-01-15"))
        XCTAssertTrue(builder === builder.dueDate("2026-02-15"))
        XCTAssertTrue(builder === builder.currency("EUR"))
        XCTAssertTrue(builder === builder.template("minimal"))
        XCTAssertTrue(builder === builder.locale("de"))
        XCTAssertTrue(builder === builder.footerText("Thanks!"))
        XCTAssertTrue(builder === builder.accentColor("#8b5cf6"))
    }

    func testBuilderValidatesRequiredFields() async throws {
        let client = try EnvoiceClient(apiKey: "env_sandbox_test")
        let result = try await client.invoice().generate()

        switch result {
        case .success:
            XCTFail("Expected failure")
        case .failure(let errors):
            XCTAssertTrue(errors.contains { $0.path == "$.invoice.number" })
            XCTAssertTrue(errors.contains { $0.path == "$.invoice.date" })
            XCTAssertTrue(errors.contains { $0.path == "$.invoice.seller" })
            XCTAssertTrue(errors.contains { $0.path == "$.invoice.buyer" })
            XCTAssertTrue(errors.contains { $0.path == "$.invoice.items" })
        }
    }

    func testBuilderAcceptsPartyObjects() throws {
        let client = try EnvoiceClient(apiKey: "env_sandbox_test")

        let seller = Party(
            name: "Acme GmbH",
            city: "Berlin",
            country: "DE",
            vatId: "DE123456789"
        )

        let builder = client.invoice().seller(seller)
        XCTAssertNotNil(builder)
    }

    func testBuilderAcceptsPartyClosure() throws {
        let client = try EnvoiceClient(apiKey: "env_sandbox_test")

        let builder = client.invoice()
            .seller { party in
                party.name = "Acme GmbH"
                party.vatId = "DE123456789"
                party.city = "Berlin"
            }

        XCTAssertNotNil(builder)
    }

    func testBuilderAcceptsLineItems() throws {
        let client = try EnvoiceClient(apiKey: "env_sandbox_test")

        let item = LineItem(
            description: "Consulting",
            quantity: 8,
            unit: "HUR",
            unitPrice: 150.0,
            vatRate: 19.0
        )

        let builder = client.invoice().addItem(item)
        XCTAssertNotNil(builder)
    }

    func testSuccessResultMethods() throws {
        let success = InvoiceSuccess(
            pdfBase64: "JVBERi0xLjQKJeLjz9MKMSAwIG9iago8PC9UeXBlL0NhdGFsb2c+PgplbmRvYmoKdHJhaWxlcgo8PC9Sb290IDEgMCBSPj4KJSVFT0YK",
            filename: "invoice-2026-001.pdf",
            validation: ValidationResult(status: "valid", profile: "EN16931", version: "2.3.2", warnings: nil),
            account: AccountInfo(remaining: 499, plan: "starter", overageCount: nil, overageAllowed: nil, warning: nil)
        )

        let data = try success.toData()
        XCTAssertTrue(data.count > 0)

        let str = String(data: data, encoding: .utf8)!
        XCTAssertTrue(str.hasPrefix("%PDF"))

        let dataUrl = success.toDataUrl()
        XCTAssertTrue(dataUrl.hasPrefix("data:application/pdf;base64,"))
    }
}
